Hw25 spring mwc test

ihavedone opened the pull request.

ihavedone committed:
a couple of tests

ihavedone committed:
fix

ihavedone committed:
spring mvc test

ihavedone committed:
Merge branch 'master' into hw25-spring-mwc-test

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/repositories/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,64 @@
+import ru.otus.merets.core.service.DBServiceUser;
+import java.util.List;
+
+@Repository
+public class UserRepositoryImpl implements UserRepository {
+    private final DBServiceUser dbServiceUser;
```
Репозиторий не может использовать сервис. Он ниже его по слою. По сути он тот же дао

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/repositories/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,64 @@
+    public UserRepositoryImpl(DBServiceUser dbServiceUser) {
+        this.dbServiceUser = dbServiceUser;
+
+        User artem = new User.Builder("Artem")
+                .withAge(30)
+                .withPhone("+70001112233")
+                .withPhone("+70001112244")
+                .withAddress("Varshavskoe highway")
+                .withPassword("123456")
+                .build();
+        User elvira = new User.Builder("Elvira")
+                .withAge(30)
+                .withAddress("Varshavskoe highway")
+                .withPassword("123456")
+                .build();
+        User dmitry = new User.Builder("Dmitry")
+                .withAge(4)
+                .withAddress("Varshavskoe highway")
+                .withPassword("123456")
+                .build();
+        User maxim = new User.Builder("Maxim")
+                .withAge(1)
+                .withAddress("Varshavskoe highway")
+                .withPassword("123456")
+                .build();
+        dbServiceUser.saveUser( artem );
+        dbServiceUser.saveUser( elvira );
+        dbServiceUser.saveUser( dmitry );
+        dbServiceUser.saveUser( maxim );
```
Инициализация БД начальными данными не обязанность конструктора. Для этого лучше сделать отдельный метод и вызвать его из конфига, при сборки, через аннотацию `@Bean`

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/repositories/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.List;
+
+@Repository
+public class UserRepositoryImpl implements UserRepository {
```
Я так понимаю этот класс нужен для инициализации. Для этого можно просто сделать еще один сервис, который использует внутри себя `DBServiceUser` и имеет один метод `intStorageWithData`. А еще лучше использовать `Flyway`

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/hibernate/dao/UserDaoHibernate.java:
```diff
@@ -0,0 +1,114 @@
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public class UserDaoHibernate implements UserDao {
```
А вот это как раз `@Repository`

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/cachehw/MyCache.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.WeakHashMap;
+
+@Component
+public class MyCache<K, V> implements HwCache<K, V> {
```
Я бы это в конфиге собрал

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/controllers/UserApiController.java:
```diff
@@ -0,0 +1,38 @@
+
+    @GetMapping("/{id}")
+    public UserDto getUserById(@PathVariable(name = "id") long id ){
+        return new UserDto(userService.getUser(id).get());
```
Лучше использовать не `get()`, а задать явно действие в случае если вернется пустой `Optional` с помощью `orElse*`

ihavedone committed:
- DI of cache is changed
- orElse in optional (instead of simple "get")
- dao -> repository
- additional helper service (init DB)

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/core/service/DBInitService.java:
```diff
@@ -0,0 +1,44 @@
+        this.dbServiceUser = dbServiceUser;
+    }
+
+    @Bean
+    public void initData(){
```
Не совсем. Имелось в виду, что будет создан конфиг (`@Configuration`), где мы определим `DBInitService` через аннотацию `@Bean` и укажем `initData` в качестве значения аргумента аннотации `initMethod`

ihavedone replied:
Кажется, готово. В новом коммите проверим. Со спрингом я пока не знаком, это все выглядит для меня не очевидно)

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/hibernate/HibernateUtils.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Arrays;
+
+@Component
+public final class HibernateUtils {
```
Лучше создавать бины в конфигурациях. Ну и классы называть соответственно. `HibernateConfig`

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/hibernate/HibernateUtils.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Bean
+    public static SessionFactory buildSessionFactory() {
```
Лучше без `static`

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/WebConfig.java:
```diff
@@ -0,0 +1,80 @@
+        return viewResolver;
+    }
+
+    @Bean
+    public MyCache<String, User> myCache(){
+        MyCache<String,User> myCache = new MyCache<>();
+        myCache.addListener((key, value, action) -> System.out.println(action+" - "+key+":"+value));
+        return myCache;
+    }
```
Не уверен, что кеш имеет отношение именно к `WebConfig`

ihavedone committed:
Merge branch 'master' into hw25-spring-mwc-test

ihavedone committed:
-cache bean has been moved into another class
-dbinitservice has been renamed into config
-some little fixes

ihavedone committed:
Merge remote-tracking branch 'origin/hw25-spring-mwc-test' into hw25-spring-mwc-test

ihavedone committed:
- created HibernateConfig using HibernateUtils
- created DBInitService for DBInitConfig
- removed some unexpected imports

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/services/DBInitServiceImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+@Service
+public class DBInitServiceImpl implements DBInitService {
+    private DBServiceUser dbServiceUser;
```
`final`

stvort commented on hw25-spring-mvc/src/main/java/ru/otus/merets/DBInitConfig.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Bean(initMethod = "initStorageWithData")
+    public DBInitService getConfig(){
```
Ага. Можно вообще прямо в метод `DBServiceUser` внедрить. Ну и метод конечно вряд ли должен называться `getConfig`

ihavedone committed:
- final
- rename one function name
